pkg/morph/deploy/nns.go

```diff
@@ -226,3 +227,174 @@ func lookupNNSDomainRecord(inv *invoker.Invoker, nnsContract util.Uint160, domai
+			// to dubious practical need
+			// Track https://github.com/nspcc-dev/neofs-contract/issues/340
+			prm.logger.Info("same local and on-chain checksums of the NNS contract NEF, update is not needed")
+			return nil
+		}
+
+		prm.logger.Info("NEF checksums of the on-chain and local NNS contracts differ, need an update")
+
+		versionOnChain, err := readContractOnChainVersion(prm.blockchain, nnsOnChainState.Hash)
+		if err != nil {
+			prm.logger.Error("failed to read on-chain version of the NNS contract, will try again later", zap.Error(err))
+			continue
+		}
+
+		// we could also try to compare on-chain version with the local one and tune
+		// update strategy: don't try to update with earlier version and catch updated
+		// contracts with unchanged version (blunder, we should react to it). For
+		// simplicity, we naively rely on the version change when the contract changes
+
+		extraUpdateArgs, err := prm.buildVersionedExtraUpdateArgs(versionOnChain)
+		if err != nil {
+			prm.logger.Error("failed to prepare build extra arguments for NNS contract update, will try again later",
+				zap.Stringer("on-chain version", versionOnChain), zap.Error(err))
+			continue
+		}
+
+		setGroupInManifest(&prm.localManifest, prm.localNEF, prm.committeeGroupKey, prm.localAcc.ScriptHash())
+
+		var vub uint32
+
+		// we pre-check 'already updated' case via MakeCall in order to not potentially
+		// wait for previously sent transaction to be expired (condition below) and
+		// immediately succeed
+		tx, err := committeeActor.MakeCall(nnsOnChainState.Hash, methodUpdate,
+			bLocalNEF, jLocalManifest, extraUpdateArgs)
+		if err != nil {
+			if isErrContractAlreadyUpdated(err) {
+				// FIXME: we get here when on-chain version is >= than the local one. If equals,
+				// this case must be considered as failure since version must be changed if code
+				// changes (NEF checksum differs here according to condition above). If greater,
+				// we should try to update at all. For simplicity, such condition is considered
+				// as success by current procedure now. To catch it, we need to pre-read version
+				// of the local contract using local NEF.
+				prm.logger.Info("NNS contract has already been updated, skip")
+				return nil
+			}
+		} else {
+			if updateTxValidUntilBlock > 0 {
+				prm.logger.Info("transaction updating NNS contract was sent earlier, checking relevance...")
+
+				if cur := prm.monitor.currentHeight(); cur <= updateTxValidUntilBlock {
+					prm.logger.Info("previously sent transaction updating NNS contract may still be relevant, will wait for the outcome",
+						zap.Uint32("current height", cur), zap.Uint32("retry after height", updateTxValidUntilBlock))
+					continue
+				}
+
+				prm.logger.Info("previously sent transaction updating NNS contract expired without side-effect")
+			}
+
+			prm.logger.Info("sending new transaction updating NNS contract...")
+
+			_, _, vub, err = committeeActor.Notarize(tx, nil)
```

Comment:
Just use the Waiter here and wait one of main/fallback to be persisted. Insufficient funds error can still be handled in this case, and you don't have to use monitor's `waitForNextBlock` to wait for every new block manually.

Reply:
we don't wanna miss successful background update from other member, that's why we `continue`. Waiter doesn't support custom conditions stopping wait loop (maybe that's an idea btw)

Reply:
> we don't wanna miss successful background update from other member

Who is responsible for NNS update? Are we supposing that several clients with different wallets will start NNS update simultaneously?

> Waiter doesn't support custom conditions stopping wait loop

It supports better solution: context that can be cancelled. See https://github.com/nspcc-dev/neo-go/blob/abb58fb7de6053dc9734b5acfa334870481b6110/pkg/rpcclient/actor/waiter.go#L57 and documentation to it.

Reply:
> It supports better solution: context that can be cancelled.

it's good, but then i should cancel context in a separate routine (which checks background side-effect). Technically i can do this, but i don't think it will improve anything really much

what do u think?

Reply:
> Who is responsible for NNS update? Are we supposing that several clients with different wallets will start NNS update simultaneously?

any committee member

Reply:
> Which way do you like?

it seems pretty good, but i will be an async notification listener, too, so `Waiter` looks easier to me if we decide to wait for side-effect asynchronously